src/python_testing/TC_EPM_2_1.py

```diff
@@ -0,0 +1,98 @@
+
+    def desc_TC_EPM_2_1(self) -> str:
+        """Returns a description of this test"""
+        return "5.1.2. [TC-EPM-2.1] Attributes with Server as DUT"
+
+    def pics_TC_EPM_2_1(self):
+        """ This function returns a list of PICS for this test case that must be True for the test to be run"""
+        return ["EPM.S"]
+
+    def steps_TC_EPM_2_1(self) -> list[TestStep]:
+        steps = [
+            TestStep("1", "Commissioning, already done", is_commissioning=True),
+            TestStep("2", "TH reads PowerMode attribute. Verify that the DUT response contains an enum8 value"),
+            TestStep("3", "TH reads NumberOfMeasurementTypes attribute. Verify that the DUT response contains an uint8 value."),
+            TestStep("4", "TH reads Accuracy attribute. Verify that the DUT response contains a list of MeasurementAccuracyStruct entries - Verify that the list has between 1 and NumberOfMeasurementTypes entries."),
+            TestStep("5", "TH reads Ranges attribute. Verify that the DUT response contains a list of MeasurementRangeStruct entries - Verify that the list has between 0 and NumberOfMeasurementTypes entries."),
+            TestStep("6", "TH reads Voltage attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of -2^62 to 2^62."),
+            TestStep("7", "TH reads ActiveCurrent attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of -2^62 to 2^62."),
+            TestStep("8", "TH reads ReactiveCurrent attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of -2^62 to 2^62."),
+            TestStep("9", "TH reads ApparentCurrent attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of 0 to 2^62."),
+            TestStep("10", "TH reads ActivePower attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of -2^62 to 2^62."),
+            TestStep("11", "TH reads ReactivePower attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of -2^62 to 2^62."),
+            TestStep("12", "TH reads ApparentPower attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of -2^62 to 2^62."),
+            TestStep("13", "TH reads RMSVoltage attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of -2^62 to 2^62."),
+            TestStep("14", "TH reads RMSCurrent attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of -2^62 to 2^62."),
+            TestStep("15", "TH reads RMSPower attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of -2^62 to 2^62."),
```

Comment:
The spec and test specs both seem wrong here, as RMS values cannot be negative. Spec and Test Specs should be being updated in parallel, so...
```suggestion
            TestStep("13", "TH reads RMSVoltage attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of 0 to 2^62."),
            TestStep("14", "TH reads RMSCurrent attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of 0 to 2^62."),
            TestStep("15", "TH reads RMSPower attribute. Verify that the DUT response contains either null or an int64 value. Value has to be between a range of 0 to 2^62."),
```

Reply:
Suggest we raise a spec issue first and then update this (and the XML and test plans).

Reply:
Again - speaking to Hasty the idea was that we could either have another attribute to indicate direction of power flow (but that could lead to race conditions) that instead allowing the sign to be negative would be a mechanism to allow power flowing back to the grid to negative.

Suggest we leave this as it is for now.